### PR TITLE
feature(table): implements table

### LINKED ISSUE #562: A11y support for tables
Investigate and come up with POC for a11y support for tables.

### PR DESCRIPTION
**Pull request by mattnolting**
Closes #562 #513 

Implements all variations of table..

Known issues:
- Styling targets generic table elements in an effort to avoid significant redundancy in markup. This approach was discussed and reviewed by several core team members before implementation. 
- `role="grid"` prevents cell focus at mobile viewport. This is noted in documentation.

### CODE REVIEW

File: src/patternfly/components/Table/docs/code.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
+## Overview
+
+Because the table component is not used for layout and presents tabular data only, it requires the use of `role="grid"`. Expandable table content (`.pf-c-table__expandable-content`) is placed within a singular `<td>` per expandable row, that can span multiple columns.
+
+## Role="grid"
+Applying `role="grid"` to tables enhances accessible interaction while in table layout, however the responsive, css grid based layout can cause unexpected interactions. Therefore, for css grid layout, it is recommended that `role="grid"` be removed.
+
+## Sortable Tables
+
+Table columns may shift when expanding/collapsing. To address this, set `.pf-m-fit-content`, or assign a width `.pf-m-width-[width]` to the corresponding `<th>`defining the column or `<td>` within the column. Width values are `[10, 15, 20, 25, 30, 35, 40, 45, 50, 60, 70, 80, 90]` or `max`.
```
Comment (michael-coker) [inline src/patternfly/components/Table/docs/code.md]: need a space or comma between "`<th>`defining".

File: src/patternfly/components/Table/docs/code.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+- One action button, positioned in the last cell of a non-expandable row.
+- Tabular data.
+- Compact presentation modifier (not compatible with expandable table).
+
+
+## Responsive layout modifiers
+
+| Class | Applied To | Outcome |
+| -- | -- | -- |
+| `.pf-m-grid-md`, `.pf-m-grid-lg`  | `.pf-c-table`   | Changes tabular layout to responsive, grid based layout as suffixed breakpoint. |
```
Comment (michael-coker) [inline src/patternfly/components/Table/docs/code.md]: `as suffixed breakpoint` - "as" should be "at"?

File: src/patternfly/components/Table/docs/table-checkboxes-actions.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+| `aria-labelledby="{title_cell_id}"` | `.pf-c-table__check`, `.pf-c-table__action` | Provides an accessible name for special cells containing interactive elements. **Required** |
+| `aria-label="[descriptive text]"` | `.pf-c-check__input`, `.pf-c-button` | Provides an accessible name for `checkboxes` and `buttons`. **Required**|
+| `id="{title_cell_id}"` | title cell `<th>` | Provides an accessible description for special cells. **Required** |
+
+
+### Usage
+
+| Class | Applied To | Outcome |
+| -- | -- | -- |
+| `.pf-c-table__check`                    | `<th>`, `<td>`                                              | Initiates a check table cell. |
```
Comment (michael-coker) [inline src/patternfly/components/Table/docs/table-checkboxes-actions.md]: there is a lot of whitespace in the "Usage" section.

File: src/patternfly/components/Table/docs/table-compact.md [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+`** All sortable table accessibility notes and usage requirements apply.`
+<br>
+`** All checkbox/action button accessibility notes and usage requirements apply.`
+
+
+### Usage
+
+| Class | Applied To | Outcome |
+| -- | -- | -- |
+| `.pf-m-compact`                           | `.pf-c-table`                                           | Modifies for a compact table. |
```
Comment (michael-coker) [inline src/patternfly/components/Table/docs/table-compact.md]: excess white space in the md table cells

File: src/patternfly/components/Table/docs/table-expandable.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+| `aria-label="[descriptive text]"` | `.pf-c-table__toggle > .pf-c-button` | Provides an accessible name for toggle button. **Required**|
+| `aria-labelledby="{title_cell_id} {button_id}"` | `.pf-c-table__toggle > .pf-c-button` | Provides an accessible description for toggle button. **Required** |
+| `id="{button_id}"` | `.pf-c-table__toggle > .pf-c-button` | Provides a reference for toggle button description. **Required** |
+
+
+### Usage
+
+| Class | Applied To | Outcome |
+| -- | -- | -- |
+| `.pf-c-table__expandable-row`             | `<tr>`                                                | Initiates an expandable row |
```
Comment (michael-coker) [inline src/patternfly/components/Table/docs/table-expandable.md]: excess white space in the md table cells

File: src/patternfly/components/Table/docs/table-simple.md [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+| -- | -- | -- |
+| `display="grid"` | `.pf-c-table` | Identifies the element that serves as the grid widget container. **Required** |
+| `aria-label` | `.pf-c-table` | Provides an accessible name for the table when a descriptive `<caption>` or `<h*>` is not available. **Required in the absence of `<caption>` or `<h*>`** |
+| `data-label="[td description]"` | `<td>` | This attribute replaces table header in mobile viewport. It is rendered by `::before` pseudo element. |
+
+### Usage
+
+| Class | Applied To | Outcome |
+| -- | -- | -- |
+| `.pf-c-table`                           | `<table>`                                           | Initiates a table element. **Required** |
```
Comment (michael-coker) [inline src/patternfly/components/Table/docs/table-simple.md]: excess white space in the md table cells

File: src/patternfly/components/Table/docs/table-sortable.md [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+| Attribute | Applied To | Outcome |
+| -- | -- | -- |
+| `aria-sort=[ascending or descending]` | `.pf-c-table__sort` | Indicates if columns in a table are sorted in ascending or descending order. For each table, authors __SHOULD__ apply aria-sort to only one header at a time. **Required**|
+
+### Usage
+
+| Class | Applied To | Outcome |
+| -- | -- | -- |
+| `.pf-c-table__sort`                     | `<th>`                                              | Initiates a sort table cell. **Required for sortable table columns** |
```
Comment (michael-coker) [inline src/patternfly/components/Table/docs/table-sortable.md]: excess white space in the md table cells

File: src/patternfly/components/Table/docs/table-checkboxes-actions.md [hunk 8]
```diff
@@ -0,0 +1,27 @@
+`** All simple table accessibility and usage requirements apply.`
+`** All sortable table accessibility notes and usage requirements apply.`
+
+Note: When including interactive elements in a table, the primary, descriptive cell in the corresponding row is a `<th>`, rather than a `<td>`. In this example, 'Node 1' and 'Node 2 siemur/test-space' are `<th>`s.
+
+
+### Thead Accessibility
+| Attribute | Applied To | Outcome |
+| -- | -- | -- |
+| `aria-label="[descriptive text]"` | `.pf-c-table__check`, `.pf-c-table__action`, `empty <th>` | Provides an accessible name for `<th>` when cell does not contain textual content. **Required when th cell does not contain textual content**|
```
Comment (michael-coker) [inline src/patternfly/components/Table/docs/table-checkboxes-actions.md]: `empty <th>` - empty shouldn't be in backticks

File: src/patternfly/components/Table/docs/table-expandable.md [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+| `.pf-c-table__expandable-row-content`     | `.pf-c-table__expandable-row > <td> > <div>`          | Modifies for expanded state. |
```
Comment (michael-coker) [inline src/patternfly/components/Table/docs/table-expandable.md]: I think the outcome text is wrong.

File: src/patternfly/components/Table/examples/index.js [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+      <Example heading="Simple Table" handlebars={tableSimpleExampleRaw} docs={tableSimpleDoc} minHeight="2em">
+        {tableSimpleExample}
+      </Example>
+      <Example heading="Sortable Table" handlebars={tableSortableExampleRaw} docs={tableSortableDoc}>
+        {tableSortableExample}
+      </Example>
+      <Example
+        heading="Table with checkboxes/actions"
+        handlebars={tableSimpleWithCheckboxesExampleRaw}
+        docs={tableCheckboxesActionsDoc}
```
Comment (michael-coker) [inline src/patternfly/components/Table/examples/index.js]: I would display these attrs on a single like.

File: src/patternfly/components/Table/styles.scss [hunk 11]
```diff
@@ -0,0 +1,792 @@
...
+  }
+}
+
+// Icon inline
+// replace this when inline text with icon is in place
+.pf-c-table__icon-inline {
+  display: flex;
+  align-items: center;
+
+  > *:not(:last-child) {
```
Comment (michael-coker) [inline src/patternfly/components/Table/styles.scss]: `*` shouldn't be necessary

File: src/patternfly/components/Table/styles.scss [hunk 12]
```diff
@@ -0,0 +1,792 @@
...
+}
+
+// Modifier - Width
+/* stylelint-disable */
+.pf-c-table [class*="pf-m-width-"] {
+  table-layout: fixed;
+}
+
+.pf-c-table .pf-m-width- {
+  &10 {
```
Comment (michael-coker) [inline src/patternfly/components/Table/styles.scss]: this breaks the rules :) I get why you're doing it, but it makes the code less searchable.

File: src/patternfly/components/Table/table-th.hbs [hunk 13]
```diff
@@ -0,0 +1,34 @@
+<th{{#if table-th--toggle}} class="pf-c-table__toggle {{table-th--modifier}}"{{else if table-th--check}} class="pf-c-table__check {{table-th--modifier}}"
+{{else if table-th--action}} class="pf-c-table__action {{table-th--modifier}}"{{else if table-th--sortable}} class="pf-c-table__sort{{#if table-th--asc}} pf-m-ascending{{/if}}{{#if table-th--desc}}pf-m-descending{{/if}} {{table-th--modifier}}"{{else if table-th--modifier}} class="{{table-th--modifier}}"{{/if}}
```
Comment (michael-coker) [inline src/patternfly/components/Table/table-th.hbs]: I think you need a space in `{{#if table-th--desc}}pf-m-descending`

File: src/site/workspace.scss [hunk 14]
```diff
@@ -270,12 +294,26 @@
...
-    border: 1px solid #ccc;
+  table:not(.pf-c-table) { 
+    td, th, tr {
+      padding: .5em;
+      vertical-align: top;
+      border: 1px solid #ccc;
+    }
+  }
+
+  .pf-c-table {
```
Comment (michael-coker) [inline src/site/workspace.scss]: should we add `margin-bottom: 0` here?

File: src/site/_site/Example/index.js [hunk 15]
```diff
@@ -109,16 +109,16 @@ export default class Example extends React.Component {
...
-            <div className="Example__documentation">
-              <p className="Example__documentation--text" dangerouslySetInnerHTML={htmlDocs} />
-            </div>
-          )}
           <div className="Example__section">
             <Preview isViewport={isViewport} heading={heading} fullPageOnly={fullPageOnly} minHeight={minHeight}>
               {children}
             </Preview>
           </div>
+          {htmlDocs && (
```
Comment (michael-coker) [inline src/site/_site/Example/index.js]: Should the order be Preview -> Handlebars -> Documentation to be consistent with the existing basic format?

File: src/patternfly/components/Table/examples/index.js (hunk 10 above)
Comment (mattnolting) [reply src/patternfly/components/Table/examples/index.js]: Agree, something in the pre-commit process is splitting this out..

**Comment by jgiardino** [comment]
These examples are really impressive! 🎆 I like how you break the examples down to highlight specific features, and provide documentation for each example. There’s so much here that’s going on that’s amazing. Great job on this one!

The responsive layouts worked for me in JAWS, with or without role grid. The positioning of elements and tab order is a _little_ wonky, but maybe that isn’t a concern in this layout? And I don’t have any great suggestions for what to do otherwise, so I’d say leave it for now until we get feedback.

I primarily reviewed this in JAWS, plus a quick review in VO. There are a couple of tweaks that I think would be good. But let me know if any of these seem questionable, or if you want any of these captured in a separate issue.

- **Empty table headers** - Can I take back a previous suggestion to include aria-label for these? I like the idea of providing context for these columns in the `<th>` row, but after testing this, the `aria-label` on the `<th>` is repeated for the table cells in that column, and it seems really redundant and noisy in JAWS. Instead, as part of a separate issue, I’d like to explore providing a description for specific table patterns, that describe the functionality available in the table.
- **Checkbox column** - The `<th>` is potentially really confusing with the labels included. The labels included here are repeated in the table body cells when navigating left-right across the table. 
    - For example, when I shift focus from Node 1 to the checkbox, I hear: “Select row column Select all rows Node 1 checkbox not checked select row column 1”
    - Can we make these changes:
        - remove aria-label from ALL the checkboxes (header row and content rows)
        - remove “column” from the `<th>` label which seems redundant
        - move `aria-labelledby` from the `<td>` to the checkbox—this will give the checkbox the label of the main `<td>` element
    - These changes would result in this: “Select row Node 1 checkbox not checked column 1”
    - For the `<th>` checkbox, in a separate issue, I’d like to add tooltip text that uses `aria-describedby` with the text “Select all rows”. This should prevent it from being repeated in the table body cells, while providing additional clarification about the function of this table header checkbox
- **Actions column** -
    - Having the aria-labelledby on the action td is redundant when the label we use is also the th for the row. What are your thoughts on just relying on the th in the row to provide context for the action? 
    - What are your thoughts about “Actions” instead of “Overflow” for the button label, to be more explicit about what the contents of the menu are? 
    - If you make these changes, we would get one of the following depending on direction travelled:
        - from adjacent column: “Actions menu button, column 7”
        - from adjacent row: “Node 1 Actions menu button, row 2”
- We chatted about using `scope="row"` for the `<th>` in the body, but while I was testing without it, I realized something that’s really nice about not having it is that with the expanded example, Node 1 is announced as a heading for the expanded section contents. We would lose that I think if we added scope in. I haven't noticed any issues with it not being included.

**Things to consider / capture as future issues:** 
- I’d like for us to consider providing an additional sr-only description about the table, that describes some of the functionality that’s available. For example, if columns are sortable, if actions are available, if rows can be selected, or expanded. I _think_ it could be beneficial to provide this context so that users have an expectation about what’s available.
- I’d like to spend some time fully testing the responsive layout across the different screen readers, and testing if role=“grid” makes a difference. As part of this issue, it would also be good to decide if we want the react implementation to make any adjustments to the HTML for the responsive layout.

File: src/patternfly/components/Table/docs/table-checkboxes-actions.md [hunk 16]
```diff
@@ -0,0 +1,26 @@
+`** All simple table accessibility and usage requirements apply.`
```
Comment (andresgalante) [inline src/patternfly/components/Table/docs/table-checkboxes-actions.md]: Should this be code? I think I'd remove the ` ` `

**Comment by andresgalante** [comment]
@mattnolting great job as always. 

When I reduce the size of the viewport there is a point where things rearrange but the labels are not bold. Is that intended?

**Comment by mattnolting** [comment]
@andresgalante The design doesn't handle the smallest breakpoint and things get very crowded in that view. So to handle I stacked the label and cell contents. @jgiardino ready for testing:)

**Comment by mattnolting** [comment]
@kybaker ready for review. take special note of portrait, mobile view

**Comment by jgiardino** [comment]
Thanks for making updates! I reviewed one more time. 

For some reason, in the example **Expandable Table**, the Node 1 cell is being announced as "blank". The only thing I can find that fixes this is by removing the `id` attribute from this `<th>` or removing the `aria-labelledby` from the checkbox. This only seems to be the case in this one instance, and not other similar instances. I tested different variations in codepen to see if there were other things that might have been causing this, but I couldn't find anything. I even added in the missing `<tr>` in the `<thead>`. The only way I could find to fix it and keep the experience as intended was to move the `id` to a `<span>`, like `<th data-label="..."><span id="...">Node 1</span></th>`. This might actually be a better solution anyway for examples like the Node 2 header that has extra text that might not be necessary as the checkbox label.

There's one change I mentioned above that you missed for the kebab menu:

- This is what I see in **Table with checkboxes and actions** 
    `<button class="pf-c-button pf-m-plain" id="simple-action1" aria-labelledby="simple-node1 simple-action1" aria-label="Actions">`
- But instead, if we do this, then we avoid the row header being announced twice (once as the 
`<th>` element for the row, and once as the `aria-labelledby` element):
    `<button class="pf-c-button pf-m-plain" id="simple-action1" aria-label="Actions">`

I forgot to mention this last time, but the second expand row toggle needs to be `aria-expanded="false"`. Also, I might have missed this or maybe it changed, but the details row for Node 2 is present in the DOM and visible to the screen reader. Was this intentional? I would suggest that the row be either not present or include the `hidden` attribute. Some of the feedback we received from our last workshop with GMS was a preference to have extra information collapsed so that it reduces the amount of contents that initially load on a page.

File: src/patternfly/components/Table/examples/table-expandable-example.hbs [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+        </div>
+      {{/table-td}}
+      {{#> table-td}}
+      {{/table-td}} 
+    {{/table-tr}}         
+  {{/table-tbody}}
+
+  {{#> table-tbody}}
+    {{#> table-tr}}
+      {{#> table-td table-td--toggle="true" table-td--button-attribute='aria-labelledby="expandable-node2 expandable-toggle2" id="expandable-toggle2" aria-label="Details" aria-expanded="true"'}}{{/table-td}}
```
Comment (jgiardino) [inline src/patternfly/components/Table/examples/table-expandable-example.hbs]: This should be `aria-expanded="false"`

File: src/patternfly/components/Table/examples/table-expandable-example.hbs [hunk 18]
```diff
@@ -0,0 +1,110 @@
+{{#> table table--grid="true" table--modifier="pf-u-box-shadow-md pf-m-grid-lg" table--expandable="true" table--attribute='aria-label="Expandable table example"'}}
+  {{#> table-thead}}
+    {{#> table-td}}{{/table-td}}
```
Comment (jgiardino) [inline src/patternfly/components/Table/examples/table-expandable-example.hbs]: The `<tr>`/`{{#> table-tr}}` is missing in this example.

**Comment by mattnolting** [comment]
@jgiardino 
> I forgot to mention this last time, but the second expand row toggle needs to be `aria-expanded="false"`. Also, I might have missed this or maybe it changed, but the details row for Node 2 is present in the DOM and visible to the screen reader. Was this intentional? I would suggest that the row be either not present or include the `hidden` attribute. Some of the feedback we received from our last workshop with GMS was a preference to have extra information collapsed so that it reduces the amount of contents that initially load on a page.

I think this is a react question. I don't think it would be difficult to reveal expandable content with a smooth transition, but REMOVING it from the DOM would impact transition. I added `hidden` to the collapsed, expandable row.

**Pull request merged**